### PR TITLE
Optimization and refactoring

### PR DESCRIPTION
**Pull request by eclique**
Hi @jacobgil 
Thanks for sharing the code. I've been working with it and here's a version I ended up with.
Please feel free to edit it and correct me if my reasoning if wrong.

Here's a list of major changes:
- At least in the current version replacing keras.relu with tf.relu is unnecessary
- There was no need in creating target_category_loss. Loss you were computing (that's y_c in the paper) can be accessed by simple indexing.
- In guided backprop there were also some strange computations with taking max and then sum. The output here should be just the gradient of conv output w.r.t. the input image.

There are some minor optimizations, such as computing GradCAM with dot product instead of the loop, accessing model's layer directly by it's name.

One difference with the paper that remains is l2-normalization in `grad_cam`. I believe authors didn't use it, but I kept it, because it helped in my case.

Modules versions: keras (2.0.8), tensorflow (1.3.0)

### CODE REVIEW

File: grad-cam.py [hunk 1]
```diff
@@ -85,59 +73,96 @@ def deprocess_image(x):
...
+    class_name = decode_predictions(np.eye(1, nb_classes, cls))[0][0][1]
+    print("Computing saliency for '{}'".format(class_name))
+
+    gradcam = grad_cam(model, preprocessed_input, cls, activation_layer=layer_name)
+    gb = guided_backprop(guided_model, img=preprocessed_input, activation_layer=layer_name)
+    guided_gradcam = gb * gradcam[..., np.newaxis]
+
+    if save:
+        jetcam = cv2.applyColorMap(np.uint8(255 * gradcam), cv2.COLORMAP_JET)
+        jetcam = (np.float32(jetcam) + np.float32(cv2.imread(sys.argv[1]))) / 2
```
Comment (23pointsNorth) [inline grad-cam.py:137]: I would put the arguments of `np.float32( cv2.imread (...) )` loading of the image as  `np.float32(image.load_img(sys.argv[1], target_size=(224, 224)))` instead.

Otherwise, it would break if the image has a different size when being loaded by the `load_image` function.
In best case scenario, the size of the image needs to be taken out as a separate parameter - e.g. top of the file.
